In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt

In [46]:
msg = pd.read_csv('datasets/messages.csv')
msg.head(4)

CONVERSATION ID CONVERSATION TITLE  \
0  2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...                NaN   
1  2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...                NaN   
2  2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...                NaN   
3  2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...                NaN   

                  FROM                                 SENDER PROFILE URL  \
0  Gaston Ariel Garcia  https://www.linkedin.com/in/gaston-ariel-garci...   
1     Juan Martin Diaz       https://www.linkedin.com/in/juan-martin-diaz   
2  Gaston Ariel Garcia  https://www.linkedin.com/in/gaston-ariel-garci...   
3  Gaston Ariel Garcia  https://www.linkedin.com/in/gaston-ariel-garci...   

                    TO                             RECIPIENT PROFILE URLS  \
0     Juan Martin Diaz       https://www.linkedin.com/in/juan-martin-diaz   
1  Gaston Ariel Garcia  https://www.linkedin.com/in/gaston-ariel-garci...   
2     Juan Martin Diaz       https://www.linkedin.com/in/juan-martin-diaz   
3     Juan Martin Diaz       https://www.linkedin.com/in/juan-martin-diaz   

                      DATE  SUBJECT  \
0  2023-08-22 21:33:15 UTC      NaN   
1  2023-08-22 13:25:29 UTC      NaN   
2  2023-08-17 15:41:07 UTC      NaN   
3  2023-08-17 15:40:47 UTC      NaN   

                                             CONTENT FOLDER  
0                   Hola . gaston.garcia@medlog.com   INBOX  
1  Gaston, te parecería que te pase esta info a a...  INBOX  
2  Perfecto,  lo tengo presente por cualquier nec...  INBOX  
3                                        Hola Juan.   INBOX

In [47]:
msg.tail(29)

CONVERSATION ID  \
2186  2-MDkxZDM4NjUtNzgyYS00NWZmLWIzOGUtM2UwMjRlZTNi...   
2187  2-MjE4NzU0ZWYtMGRjZS00YTliLWFiYWItNzdhNmNlNjFh...   
2188  2-ZjViNDBmOWYtMmE0ZC00NWZmLTkwNzYtODIzNWVhZjRm...   
2189  2-MWYxMTc3OWMtYTJlZS00MjM5LWI5NDAtNTg2ZWM3YzEz...   
2190  2-YWQ4ZjAzMzAtM2JjMC00YWZkLWFkMzMtNDJiNWNhYjBm...   
2191  2-NGI3ODJlOWMtZWQ2Yi00NzM4LWE3ODItOTg2YWY1ODZi...   
2192  2-NzZhMjI3ZDItYzJmZS00OThiLWI0NzUtZDQ3ZTFiN2I2...   
2193  2-NGNjNWVkOGItZThiZC00ZWI4LTljYmUtMmFmZWM2NzE3...   
2194  2-MTkzMzQ2YmMtNjQ4YS00ZWVlLWE0MmMtMDIxOTAyNjBm...   
2195  2-ZjRhYmQ4OWQtNDI5Ni00NjM5LWJmZWMtZjVhNjM0NzVm...   
2196  2-MzFjNzc5ZmMtNDg0Ny00ZjlmLWE5YzMtNDM4YmJkOTUx...   
2197  2-NTRiNmE3OTEtNzQ5MC00NjlkLWFmMjYtODc0ZmM4ZGVj...   
2198  2-MWJmM2NkMTUtZWVlYS00MGVhLTlmMGMtZjQ0OTkyZWI1...   
2199  2-ODI0ZjViZDItNjUyZS00Mjk4LTkxYzMtNzYyNTk0MGQ1...   
2200  2-NzRiOTRkMDQtNjU3MS00YjgwLWE1ODYtYmI2MzlmY2Nm...   
2201  2-NGY4NWM3MzktNzY5Ny00MjVlLWI5MmEtOWI3NmI4ODBj...   
2202  2-NGZkYjIzYTgtZTJkYS00MTdlLWEyYzUtZWM1NDhiMGNh...   
2203  2-YTIzNWEwZjQtY2I5My00N2M0LThjZWYtMThjNDBmMzE0...   
2204  2-NmQ2ODMyMTQtYTM3My00NzI2LTgwMTQtZjVkYmEzODAz...   
2205  2-NzY5ZmU4ZmMtZjc4ZS00ZGI4LTk0OTctMTNhMDc2YmU3...   
2206  2-MDhiYmI5ZGEtMTU3Yy00MjYyLTk1NGItMjUxYWEyOGJk...   
2207  2-ZWRhYzA5MzAtNGM1Yy00MDg3LWI5OWUtZGNhYzMxYWYz...   
2208  2-ZWRhYzA5MzAtNGM1Yy00MDg3LWI5OWUtZGNhYzMxYWYz...   
2209  2-ZWRhYzA5MzAtNGM1Yy00MDg3LWI5OWUtZGNhYzMxYWYz...   
2210  2-YzJlYjUzNzMtZmYwMC00NjNiLWIxNmQtY2U3OGI0ZDU5...   
2211  2-ODdiYzQyYWUtZjQ3OC00N2I3LWIyZmYtZDU2MzYzMGE0...   
2212  2-YTFmMjdkYjAtMGFmMi00MjNkLWE0NTMtZTVmNjNhZWU0...   
2213  2-MjM1NGYwM2MtOTk3Mi01MDg0LTg1YzYtZmIyZGUzZTgx...   
2214  2-ZmU5NjgxNWUtYzVmZS01YTA3LThmMzctMTUyODM3ZTky...   

                                     CONVERSATION TITLE  \
2186                                                NaN   
2187                             Sponsored Conversation   
2188                             Sponsored Conversation   
2189                             Sponsored Conversation   
2190                             Sponsored Conversation   
2191            Baldor SRL "Laboratorio de Calibración"   
2192                             Sponsored Conversation   
2193                             Sponsored Conversation   
2194  La campaña de Volkswagen que los condujo al éxito   
2195   Conocé la exitosa estrategia regional de L’Oréal   
2196                             Sponsored Conversation   
2197              Baldor SRL "Instrumentos de Medición"   
2198                                ¿Conoces SAP RISE?    
2199          El universo BTL necesitaba algo distinto.   
2200            Baldor SRL "Laboratorio de Calibración"   
2201                  El impacto que tu marca necesita.   
2202          El universo BTL necesitaba algo distinto.   
2203                Gracias por ser usuario de LinkedIn   
2204           Invertí en Real Estate en Estados Unidos   
2205  Prueba LinkedIn Sales Navigator gratis durante...   
2206           Invertí en Real Estate en Estados Unidos   
2207                             Sponsored Conversation   
2208                             Sponsored Conversation   
2209                             Sponsored Conversation   
2210                Gracias por ser usuario de LinkedIn   
2211                            Prueba LinkedIn Premium   
2212                Gracias por ser usuario de LinkedIn   
2213                                                NaN   
2214                                                NaN   

                                     FROM  \
2186                     Juan Martin Diaz   
2187                      LinkedIn Member   
2188                      LinkedIn Member   
2189                      LinkedIn Member   
2190                      LinkedIn Member   
2191                      Fernando Dolber   
2192                      LinkedIn Member   
2193                      LinkedIn Member   
2194                          Mercado Ads   
2195                          Mercado Ads   
219

Viendo algunos de los ultimos registros de nuestra base de datos, podemos concluir que allí se situan mensajes que consideraremos como Spam, ya que provienen de cuentas de Linkedin Premium, Linkedin Learning y otras cuentas, las que no aportan nada a nuestro analisis.

Borraremos estos registros, que son los ultimos 28, es decir, desde el index 2187 al 2214

In [48]:
registros_a_eliminar = [i for i in range(2187,2215)]
msg.drop(registros_a_eliminar, inplace=True)

Mirandolo mejor, me estoy dando cuenta que los datos estan ordenados por fecha de forma descendente

In [49]:
msg.shape
# es decir, tenemos un total de 2215 mensajes recibidos/enviados en nuestra casilla 

(2187, 10)

In [50]:
msg.columns

Index(['CONVERSATION ID', 'CONVERSATION TITLE', 'FROM', 'SENDER PROFILE URL',
       'TO', 'RECIPIENT PROFILE URLS', 'DATE', 'SUBJECT', 'CONTENT', 'FOLDER'],
      dtype='object')

In [51]:
msg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2187 entries, 0 to 2186
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   CONVERSATION ID         2187 non-null   object 
 1   CONVERSATION TITLE      2 non-null      object 
 2   FROM                    2180 non-null   object 
 3   SENDER PROFILE URL      2178 non-null   object 
 4   TO                      2179 non-null   object 
 5   RECIPIENT PROFILE URLS  2179 non-null   object 
 6   DATE                    2187 non-null   object 
 7   SUBJECT                 0 non-null      float64
 8   CONTENT                 2150 non-null   object 
 9   FOLDER                  2187 non-null   object 
dtypes: float64(1), object(9)
memory usage: 171.0+ KB


In [52]:
# removeremos las columnas que no nos sirven
msg.drop(columns=['CONVERSATION TITLE', 'SUBJECT', 'SENDER PROFILE URL', 'RECIPIENT PROFILE URLS'], inplace=True)
msg

CONVERSATION ID                 FROM  \
0     2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...  Gaston Ariel Garcia   
1     2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...     Juan Martin Diaz   
2     2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...  Gaston Ariel Garcia   
3     2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...  Gaston Ariel Garcia   
4     2-MGIyMDY4NDMtYjU1YS00ODY3LTkyMjQtY2I5MTMxZWIw...     Juan Martin Diaz   
...                                                 ...                  ...   
2182  2-NjRkYjhmNjUtNmRmNC00YThhLWJmNjQtNDQ0Y2FiZjVm...          Ruben Lopez   
2183  2-NjRkYjhmNjUtNmRmNC00YThhLWJmNjQtNDQ0Y2FiZjVm...     Juan Martin Diaz   
2184  2-MTI2ZDU5YmItNTE3NC00NjNjLThmNzktYmRiNDlkZDAx...     Juan Martin Diaz   
2185  2-MTI2ZDU5YmItNTE3NC00NjNjLThmNzktYmRiNDlkZDAx...     Fernando Aguirre   
2186  2-MDkxZDM4NjUtNzgyYS00NWZmLWIzOGUtM2UwMjRlZTNi...     Juan Martin Diaz   

                       TO                     DATE  \
0        Juan Martin Diaz  2023-08-22 21:33:15 UTC   
1     Gaston Ariel Garcia  2023-08-22 13:25:29 UTC   
2        Juan Martin Diaz  2023-08-17 15:41:07 UTC   
3        Juan Martin Diaz  2023-08-17 15:40:47 UTC   
4     Gaston Ariel Garcia  2023-08-17 12:44:50 UTC   
...                   ...                      ...   
2182     Juan Martin Diaz  2021-12-18 09:54:35 UTC   
2183          Ruben Lopez  2021-12-14 02:51:34 UTC   
2184     Fernando Aguirre  2021-12-07 21:02:22 UTC   
2185     Juan Martin Diaz  2021-12-07 17:35:03 UTC   
2186       Camila Fonseca  2023-07-19 17:56:46 UTC   

                                                CONTENT   FOLDER  
0                      Hola . gaston.garcia@medlog.com     INBOX  
1     Gaston, te parecería que te pase esta info a a...    INBOX  
2     Perfecto,  lo tengo presente por cualquier nec...    INBOX  
3                                           Hola Juan.     INBOX  
4     Buen día Gaston, como estas? Un gusto conectar...    INBOX  
...                                                 ...      ...  
2182                                            Gracias    INBOX  
2183        ¡Felicidades por tu aniversario de trabajo!    INBOX  
2184                                            Gracias    INBOX  
2185                  ¡Felicidades por tu nuevo puesto!    INBOX  
2186  Camila como estas? Un gusto conectar.  Me post...  ARCHIVE  

[2187 rows x 6 columns]

In [53]:
msg['FROM'].value_counts()


FROM
Juan Martin Diaz            1392
Nicolás Coria                 29
Maria Laura Gauna Fredes      22
Joaquin Gamez Maggiori        21
Juan cruz Vittori             16
                            ... 
Kevin Olguin                   1
Matías Garciandia              1
Carolina Funes                 1
Daiana Turchi                  1
Fernando Aguirre               1
Name: count, Length: 242, dtype: int64

In [54]:
msg['DATE'] = pd.to_datetime(msg['DATE']) # transformamos a tipo fecha
msg['DATE'] = msg['DATE'].dt.strftime(date_format='%Y-%m-%d %H:%M:%S') # quitamos el +00:00 de la zona horaria
msg['DATE']

0       2023-08-22 21:33:15
1       2023-08-22 13:25:29
2       2023-08-17 15:41:07
3       2023-08-17 15:40:47
4       2023-08-17 12:44:50
               ...         
2182    2021-12-18 09:54:35
2183    2021-12-14 02:51:34
2184    2021-12-07 21:02:22
2185    2021-12-07 17:35:03
2186    2023-07-19 17:56:46
Name: DATE, Length: 2187, dtype: object

La cuenta de Linkedin puede tener muchos mensajes que no son de utilidad para nuestro analisis, la fecha de analisis es algo que se debera tener en cuenta. 

Esto es así porque por ejemplo, en el año 2021 la empresa no siquiera estaba realizando esta campaña de marketing, por eso con el objetivo de mejorar la calidad de los datos vamos a fijar una fecha para analizar los datos de ahi en adelante
- Fecha de comienzo de campaña: 2022-04-01

In [63]:
msg = msg.loc[msg['DATE']>'2022-04-01'].reset_index(drop=True)


In [65]:
len(msg) # notar que pasamos de tener 2215 registros a tener 2099

2099

El siguiente paso seria crear una api con fastapi y render para mostrar en vivo un analisis de los datos.